src/lib.rs

```diff
@@ -16,15 +17,45 @@ use uuid::Uuid;
 pub struct ProxyCommand {
     #[serde(flatten)]
     command: Command,
+    timestamp: Option<u64>,
     signature: Option<Signature>,
 }
 
 impl ProxyCommand {
     fn verify_signature(&self, verifying_key: &Option<VerifyingKey>) -> bool {
         match (verifying_key, &self.signature) {
             (Some(key), Some(signature)) => {
-                let message = serde_json::to_string(&self.command).unwrap();
-                key.verify(message.as_bytes(), signature).is_ok()
+                let mut message = serde_json::to_string(&self.command).unwrap();
+
+                let timestamp = if let Some(timestamp) = self.timestamp {
+                    message.push_str(&timestamp.to_string());
+                    time::Duration::from_secs(timestamp)
+                } else {
+                    tracing::debug!("timestamp missing while signature is present");
+                    return false; // timestamp missing with signature present
+                };
+
+                if !key.verify(message.as_bytes(), signature).is_ok() {
+                    tracing::debug!("signature does not match message");
+                    return false; // signature doesn't match
+                }
+
+                let now = time::SystemTime::now()
+                    .duration_since(time::UNIX_EPOCH)
+                    .unwrap();
+                if now > timestamp {
```

Comment:
leave some wiggle room for commands being from the future, computer clocks not always synced. also reducing nested if
old:
```rust
                if now > timestamp {
                    // timestamp is older
                    if now - timestamp <= time::Duration::from_secs(60) {
                        // less than a minute old
                        true
                    } else {
                        tracing::warn!("command is more than a minute old");
                        false
                    }
                } else {
                    tracing::warn!("command is from the future");
                    false
                }
```
new:
```rust
                if  timestamp > (now + time::Duration::from_secs(30)) {
                    tracing::warn!("command is more than 30s from the future");
                    false
                }
                // timestamp is older
                if now - timestamp <= time::Duration::from_secs(60) {
                    // less than a minute old
                    true
                } else {
                    tracing::warn!("command is more than a minute old");
                    false
                }
```